packages/typescript-estree/src/useProgramFromProjectService.ts

```diff
@@ -0,0 +1,34 @@
+import type * as ts from 'typescript';
+import type { server } from 'typescript/lib/tsserverlibrary';
+
+import { createProjectProgram } from './create-program/createProjectProgram';
+import type { ASTAndDefiniteProgram } from './create-program/shared';
+import type { MutableParseSettings } from './parseSettings';
+
+export function useProgramFromProjectService(
+  projectService: server.ProjectService,
+  parseSettings: Readonly<MutableParseSettings>,
+): ASTAndDefiniteProgram | undefined {
+  projectService.setCompilerOptionsForInferredProjects({
+    rootDir: parseSettings.tsconfigRootDir,
+  });
+
+  projectService.openClientFile(
+    parseSettings.filePath,
+    parseSettings.codeFullText,
+    /* scriptKind */ undefined,
+    parseSettings.tsconfigRootDir,
+  );
```

Comment:
We're seeing unit test failures like `Cannot read properties of undefined (reading 'fileName')` on `sourceFile.fileName`:

```js
// (code simplified from node_modules/tslint/lib/linter.js)
var sourceFile = this.program.getSourceFile(fileName);

// ...

if (error_1.isError(error) && error.stack !== undefined) {
    error_1.showRuleCrashWarning(error.stack, rule.getOptions().ruleName, sourceFile.fileName);
}
```

That `sourceFile` is `undefined` because the project (`this.program`) created by the project service has a different `currentDirectory` than our `parseSettings.tsconfigRootDir`:

```diff
- /Users/josh/repos/typescript-eslint/packages/eslint-plugin-tslint
+ /Users/josh/repos/typescript-eslint/packages/eslint-plugin-tslint/tests/fixtures/fixture-project
```

It comes from `ConfiguredProject`'s constructor providing `currentDirectory` as `getDirectoryPath(configFileName)`:

https://github.com/microsoft/TypeScript/blob/df3bec66748f48badad61faa3fb4f9a79a11a612/src/server/project.ts#L2600

@bradzacher @jakebailey I'm not sure whether we're holding this wrong. I _think_ we would want to be able to resolve the client file using its path relative to typescript-eslint's configured root dir is, not relative to its specific project, no?

Reply:
I would think that if using ProjectService, you should totally ignore any and all other configuration knobs in ts-eslint for the project; they're not going to be used at all by tsserver. The paths you give it should be in whatever format tsserver requires (absolute? normalized? some combo?).